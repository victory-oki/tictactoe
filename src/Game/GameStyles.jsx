import styled from 'styled-components';
import bg from '../assets/bg.jpg';
import { white, black, orange, red } from '../theme/colors';


export const GameContainerStyled = styled.div`
  width: 100%;
  min-height: 100vh;
  padding: 0.1px;
  position: relative;
  background-color: ${black[100]};
  &::before {
    content: "";
    background-image: url(${bg});
    background-size: contain;
    position: absolute;
    top: 0px;
    right: 0px;
    bottom: 0px;
    left: 0px;
    opacity: 0.05;
    z-index: 1;
  }
  .container{
    max-width: 112rem;
    margin: 0rem auto;
    position: relative;
    z-index: 2;
  }
  .nav{
    padding: 1rem calc((100vw - 112rem) / 2);
    background: ${black[100]};
    width: 100vw;
    box-shadow: 5px 5px 15px 5px rgba(${black[100]});
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: fixed;
    top: 0;
    left: 0;

    .title{
      font-size: 2.4rem;
      color: ${white[100]};
      font-weight: bold;
    }
    .list{
      list-style: none;
      display: flex;
      & >:not(:last-child){
        margin-right: 4rem;
      }
    }
    &__icon{
      font-size: 2rem;
      cursor: pointer;
      color: ${white[100]};
    }
  }
  .header{
    margin: 7rem 3rem;
    text-align: center;
    &__title{
      font-size: 5.4rem;
      color: ${white[100]};
      margin-bottom: 2.5rem;
      font-weight: bold;
    }
    &__brief{
      font-size: 2rem;
      color: ${white[200]};
      font-weight: 500;
      display: flex;
      align-items: center;
      justify-content: space-between;
      max-width: 40rem;
      margin: 2rem auto;
    }
  }
  .item{
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    & >:not(:last-child){
        margin-bottom: 2rem;
    }
    & >:first-child{
        font-family: 'Fredoka One', cursive;
        fill: ${white[100]};
        width: 2.4rem;
    }
    &--1{
      color: ${orange[100]};
    }
    &--2{
      color: ${orange[200]};
    }
  }
  .status{
    background: ${black[200]};
    padding: 2.4rem;
    display: flex;
    align-items: flex-end;
    border-radius: 1.9rem;
    max-width: 50rem;
    margin: 0 auto;
    margin-bottom: 2.5rem;
    &__icon{
      font-size: 4.5rem;
      margin-right: 2rem;
    }
    &__text{
      color: ${white[100]};
      width: 100%;
      &--heading{
        font-size: 1.6rem;
        font-weight: bold;
      }
      &--content{
        font-size: 2.4rem;
        font-weight: 400;
      }
    }
    &__footer{
      display: flex;
      align-items: center;
      justify-content: space-between;
    }
    &__btn{
      color: #d3d3d3 !important;
      font-size: 1.6rem;
    }
  }
  .game{
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  .brief{
    font-size: 2rem;
    color: ${white[200]};
    font-weight: 500;
  }
  .content{
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 2rem 0;
    margin-bottom: 6rem;
  }

  .player{
    background: ${black[200]};
      width: 17rem;
      height: 18rem;
      border-radius: 2rem;
      display: flex;
      flex-direction: column;
      align-items: center;
      padding: 1.2rem;

      &__img{
        display: flex;
        width: 8rem;
        height: 8rem;
        border-radius: 50%;
        border: 0.5rem;
        background-color: grey;
        margin-top: -5rem;
        text-align: center;
        margin-bottom: 2rem;
        border: 3px solid ${white[100]};
        position: relative;
        .pi{
          position: absolute;
          display: none;
          color: ${white[100]};
          font-size: 2.5rem;
          top: -50%;
          left: 30%;
          /* transform: translate(-125%,0); */
          &.show{
            display: inline-block;
          }
        }
      }
      &__id{
        font-size: 1.2rem;
        font-weight: 200;
        color: ${white[100]};
        text-align: center;
      }
      &__name{
        font-size: 1.6rem;
        color: ${white[100]};
        text-align: center;
      }
      &__stats{
        display: flex;
        align-items: center;
        justify-content: space-between;
        margin-top: 1rem;
        width: 100%;
        font-size: 1.6rem;
        font-weight: 400;
        color: ${white[200]};
        margin-top: auto;
        &--win{}
        &--draw{
        }
      }
      &__char{
        font-family: 'Fredoka One', cursive;
        font-size: 2rem;
        color: ${white[100]};
        position: absolute;
        bottom: -1rem;
        right: 0;
        width: 3rem;
        height: 3rem;
        border-radius: 50%;
        background-color: ${white[50]};
        display: flex;
        align-items: center;
        justify-content: center;
        &--right{
          left: 0;
          right: unset;
        }
      }
  }
  .grid{
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 50rem;
      height: 50rem;
      /* border: 1px solid #515151; */
      margin: 0 auto;
      & >:not(:last-child){
        margin-bottom: 2rem;
      }
      .row{
          width: 100%;
          display: flex;
          flex: 1;
          & >:not(:last-child){
            margin-right: 2rem;
          }
      }
      .col{
          width: 100%;
          height: 100%;
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: 1.6rem;
          cursor: pointer;
          border-radius: 1.5rem;
          transition: all .4s ease-in-out;
          font-size: 4rem;
          font-family: 'Fredoka One', cursive;
          background: hsla(0, 0%, 98%, 1);
          background: radial-gradient(circle, hsla(0, 0%, 98%, 1) 100%, hsla(177, 74%, 88%, 1) 100%);
          background: -moz-radial-gradient(circle, hsla(0, 0%, 98%, 1) 100%, hsla(177, 74%, 88%, 1) 100%);
          background: -webkit-radial-gradient(circle, hsla(0, 0%, 98%, 1) 100%, hsla(177, 74%, 88%, 1) 100%);

          &:hover{
            background: hsla(0, 0%, 98%, 1);
            background: radial-gradient(circle, hsla(0, 0%, 98%, 1) 56%, hsla(177, 74%, 88%, 1) 100%);
            background: -moz-radial-gradient(circle, hsla(0, 0%, 98%, 1) 56%, hsla(177, 74%, 88%, 1) 100%);
            background: -webkit-radial-gradient(circle, hsla(0, 0%, 98%, 1) 56%, hsla(177, 74%, 88%, 1) 100%);
          }
      }
  }
  .time-label{
    font-size: 2.4rem;
    font-weight: 500;
    &--warning{
      color: ${orange[100]};
    }
    &--danger{
      color: ${red[100]};
    }
  }
`
