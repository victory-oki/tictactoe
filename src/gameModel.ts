export interface IjoinObj {
    actionType: 'CREATE' | 'JOIN';
    username: string;
}

export interface IPlayerData {
    name: string,
    playerNo: number,
    character: E_GAMECHARACTER,
    wins: number,
    losses: number,
    draws: number,
    imageId: number
}

export interface IGameData {
    playersData: IPlayerData[]
    playCount: number,
    gridDimension: number,
    currentPlayLocation: number[],
    gameGrid: string,
    gameMode:E_GAMEMODE
}

export interface IUpdateGameData {
    playLocation: number[]
    playCount: number,
    gameGrid?: string,
    roomId?: string
}

export interface IPlayAgainData {
    roomId: string,
    isDraw: boolean,
    winCharacter: string
}

export interface IGameStats {
    xwins: number,
    owins: number,
    draws: number
}
export interface IGameState {
    playerName: string,
    playerNo: number,
    playerImage?: number,
    friendNo: number,
    friendName: string,
    friendImage?: number,
    playerCharacter: E_GAMECHARACTER | string
    friendCharacter: string,
    wins: number,
    losses: number,
    draws: number,
    playCount: number,
    gridDimension: number,
    gameGrid: string,
    gameMode: E_GAMEMODE,
    currentPlayLocation: number[]
}

export const enum GameStatus {
    InPlay = 'InPlay',
    Win = 'Win',
    Draw = 'Draw'
}

export type GridCol = string | null;

export interface IGameProps {
    gameStatus: GameStatus;
    msgStatus: string;
    winLocation: Set<string>;
    grid: GridCol[][],
    winCharacter: string
}

export const EVENTS = {
    CREATE_SUCCESS: 'create_success',
    JOIN_SUCCESS: 'join_success',
    JOIN: 'join',
    ERROR: 'error',
    GAME_START: 'game_start',
    PLAY_AGAIN: 'play_again',
    GAME_RESTART_AGAIN: 'game_restart_again',
    GAME_CHECK: 'game_check',
    GAME_CHECK_SUCCESS: 'game_check_success',
    GAME_CHECK_FAILED: 'game_check_failed',
    GAME_UPDATE: 'game_update',
    SETTINGS_UPDATE: 'settings_update',
    SETTINGS_CHANGED: 'settings_changed',
}

export interface IGameHooks {
    gameReady: boolean,
    isFullscreen: boolean,
    game: IGameProps,
    playerImage: number | undefined,
    playerNo: number,
    playerName: string,
    playerCharacter: string,
    friendImage: number | undefined,
    friendNo: number,
    friendName: string,
    friendCharacter: string,
    seconds: number,
    gameMode: E_GAMEMODE,
    gridDimension:number,
    wins: number,
    losses: number,
    draws: number,
    playCount: number,
    StatusIcon: { [key in GameStatus]: string },
    renderCharacterColor: (value: string) => string,
    canPlay: () => boolean,
    isWin: () => boolean,
    isDraw: () => boolean,
    isPlayLocation: (row: number, col: number) => boolean,
    isWinLocation: (row: number, col: number) => boolean,
    playTurn: (row: number, col: number) => void,
    playAgain: () => void,
    toggleFullScreen: () => void,
}


export const enum E_GAMEMODE {
    NORMAL = 'NORMAL',
    BLITZ = 'BLITZ'
}

export const enum E_GAMECHARACTER {
    X = 'X',
    O = 'O'
}
