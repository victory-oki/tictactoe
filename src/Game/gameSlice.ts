import { createSlice } from "@reduxjs/toolkit";
import { E_GAMEMODE, IGameData, IGameState, IPlayerData, IUpdateGameData } from "../gameModel";

const initialState: IGameState = {
    playerName: '',
    playerNo: 0,
    friendNo: 0,
    friendName: '',
    playerCharacter: '',
    friendCharacter: '',
    playerImage: undefined,
    friendImage: undefined,
    wins: 0,
    losses: 0,
    draws: 0,
    playCount: 0,
    gameGrid: '',
    gameMode: E_GAMEMODE.NORMAL,
    gridDimension:3,
    currentPlayLocation: []
}

const gameSlice = createSlice({
    name: 'game',
    initialState,
    reducers: {
        setPlayerName: (state, action) => {
            const payload = action.payload;
            state.playerName = payload;
        },
        initializeGame: (state, action: { payload: IGameData }) => {
            const players: IPlayerData[] = action.payload.playersData;
            players.forEach(element => {
                if (state.playerName === element.name) {
                    state.playerCharacter = element.character;
                    state.wins = element.wins;
                    state.losses = element.losses;
                    state.draws = element.draws;
                    state.playerNo = element.playerNo;
                    state.playerImage =  element.imageId;
                }
                else {
                    state.friendName = element.name;
                    state.friendCharacter = element.character;
                    state.friendNo = element.playerNo;
                    state.friendImage = element.imageId
                }
            });
            state.currentPlayLocation = action.payload.currentPlayLocation;
            state.playCount = action.payload.playCount;
            state.gameGrid = action.payload.gameGrid;
            state.gameMode = action.payload.gameMode;
            state.gridDimension = action.payload.gridDimension;
        },
        updateGame: (state, action: { payload: IUpdateGameData }) =>{
            const payload = action.payload
            if(payload.gameGrid){
                state.gameGrid = payload.gameGrid
            }
            state.playCount = payload.playCount
            state.currentPlayLocation = payload.playLocation;
        }
    }
})

export const { setPlayerName, initializeGame, updateGame } = gameSlice.actions;
export default gameSlice.reducer
