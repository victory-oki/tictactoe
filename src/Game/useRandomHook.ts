import { useState } from "react"
import { GridCol } from "../gameModel";

const useRandomHook = (gridDimension:number):any => {
    const generatePlayLocations = (grid = null):Map<number, number[]>=>{
        const mapLength = gridDimension * gridDimension;
        const locations = new Map();
        let row = 0;
        for(let i=0; i< mapLength; i++){
            if(i%gridDimension === 0 && i !== 0){
                row += 1;
            }
            const col = i - (row * gridDimension);
            if(grid && grid[row][col]) continue;
            locations.set(i, [row, col]);
        }
        return locations;
    }

    const [playLocations, setPlayLocations]  =  useState(generatePlayLocations());
    const invalidateLocation = (index:number)=>{
        console.log('Invalidating... size before: ', playLocations.size);
        playLocations.delete(index);
        console.log('Invalidated... size after: ', playLocations.size);
    }
    const findRandomLocation = ():number[]=>{
        const randomIndex = Math.floor(Math.random() * playLocations.size);
        return Array.from(playLocations.values())[randomIndex];
    }
    const reBuildPlayLocation = (grid:GridCol[][])=>{
        setPlayLocations(generatePlayLocations(grid as any))
    }
    return {
        invalidateLocation,
        findRandomLocation,
        reBuildPlayLocation
    }
}

export default useRandomHook;
