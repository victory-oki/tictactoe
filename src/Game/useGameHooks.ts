import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { NavigateFunction, useNavigate, useParams } from "react-router-dom";
import { EVENTS, E_GAMECHARACTER, E_GAMEMODE, GameStatus, GridCol, IGameData, IGameHooks, IGameProps, IGameStats, IPlayAgainData, IPlayerData, IUpdateGameData } from "../gameModel";
import { useToast } from "../notificationContext";
import { useSocket } from "../socketContext";
import { RootState } from "../store";
import { initializeGame, updateGame } from "./gameSlice";
import * as GameUtils from '../gameutils';
import { orange } from '../theme';
import useRandomHook from "./useRandomHook";
import { useAudio } from "../components/SoundPlayer/useAudioHooks";
import { defaultConfig } from "../components/SoundPlayer/SoundPlayer";

const useGameHooks = (): IGameHooks => {
  const gridInit: GridCol[][] = [[]];
  const GRID_DIMENSION = 3;
  const [gameReady, setGameReady] = useState<boolean>(false);
  const [seconds, setSeconds] = useState(59);
  const [intervalId, setIntervalId] = useState<any>(0);
  const [isFullscreen, setIsFullscreen] = useState<boolean>(false);
  const [game, setGame] = useState<IGameProps>({
    gameStatus: GameStatus.InPlay,
    msgStatus: '',
    winLocation: new Set(),
    grid: gridInit,
    winCharacter: ''
  });
  const navigate: NavigateFunction = useNavigate();
  const { id: gameId } = useParams();
  const socket = useSocket();
  const dispatch = useDispatch();
  const toast = useToast();
  const {audio:playSound} = useAudio(require('../assets/sound2.wav'), defaultConfig);
  const {
    playerName,
    friendName,
    playerNo,
    friendNo,
    playerCharacter,
    friendCharacter,
    gameGrid,
    playCount,
    playerImage,
    friendImage,
    gameMode,
    wins,
    losses,
    draws,
    currentPlayLocation,
    gridDimension
  } = useSelector((store: RootState) => store.game);
  const { invalidateLocation, findRandomLocation, reBuildPlayLocation } = useRandomHook(gridDimension);
  useEffect(() => {
    setGameIfExist();
    setupSocketListeners();
    const fullScreenChangeHandler = ()=>{
      const isFullscreen = document.fullscreenElement? true: false;
      setIsFullscreen(isFullscreen);
    }
    document.addEventListener('fullscreenchange', fullScreenChangeHandler);
    return () => {
      destroyListeners();
      document.removeEventListener('fullscreenchange',fullScreenChangeHandler )
    }
  }, [gameId]);

  useEffect(() => {
    if (gameGrid) {
      let status = GameStatus.InPlay;
      let winCharacter = '';
      const newSet: Set<string> = new Set();
      const newGrid = JSON.parse(gameGrid);

      if (gameMode === E_GAMEMODE.BLITZ && currentPlayLocation.length > 0) {
        const [row, col] = currentPlayLocation;
        const index = row * GRID_DIMENSION + col
        invalidateLocation(index);
      }

      if (playCount > 2 && currentPlayLocation.length > 0) {
        const [row, col] = currentPlayLocation;
        const winLocation = getWinLocationIfExists(row, col, newGrid);
        if (winLocation.length > 0) {
          winCharacter = newGrid[winLocation[0][0]][winLocation[0][1]];
          winLocation.forEach(location => newSet.add(`${location[0]}${location[1]}`));
          status = GameStatus.Win;
          stopCountDown();
        }
        else if (playCount === Math.pow(gridDimension,2)) {
          status = GameStatus.Draw;
          stopCountDown();
        }
      }
      const game: IGameProps = {
        gameStatus: status,
        msgStatus: updateStatusMessage(status, winCharacter),
        winLocation: newSet,
        grid: newGrid,
        winCharacter
      }
      setGame(game);
    }
  }, [gameGrid, playerCharacter, gameMode]);
  useEffect(() => {
    if (playCount && gameMode === E_GAMEMODE.BLITZ && canPlay() && !isWin() && !isDraw()) {
      startCountDown();
    }
    else {
      stopCountDown();
    }
  }, [playCount])
  useEffect(() => {
    if (seconds === 0) {
      stopCountDown();
      const msg = 'You ran out of time, A random spot was played for you';
      toast.current?.show({ severity: 'info', summary: 'Play Again', detail: msg, life: 3000 });
      const [row, col] = findRandomLocation();
      playTurn(row, col);
    }
  }, [seconds])
  const setupSocketListeners = (): void => {
    socket.on(EVENTS.GAME_CHECK_SUCCESS, (gameData: IGameData) => {
      if (isUserNameSet()) {
        dispatch(initializeGame(gameData));
        setGameReady(true);
        if (gameData.gameMode === E_GAMEMODE.BLITZ) {
          const grid = JSON.parse(gameData.gameGrid);
          reBuildPlayLocation(grid);
        }
      }
      else {
        navigate(`/join?gameId=${gameId}`, { replace: false });
      }
    })
    socket.on(EVENTS.GAME_UPDATE, (updateData: IUpdateGameData) => {
      dispatch(updateGame(updateData))
    })
    socket.on(EVENTS.GAME_CHECK_FAILED, (msg: string) => {
      navigate(`/create`, { replace: false, state: msg })
    })
    socket.on(EVENTS.GAME_RESTART_AGAIN, (gameData: IGameData) => {
      const msg = 'Game has started Again, Enjoy 🎉'
      toast.current?.show({ severity: 'info', summary: 'Play Again', detail: msg, life: 3000 });
      dispatch(initializeGame(gameData));

      if (gameData.gameMode === E_GAMEMODE.BLITZ) {
        const grid = JSON.parse(gameData.gameGrid);
        reBuildPlayLocation(grid);
      }
    })
    socket.on(EVENTS.SETTINGS_CHANGED, (gameData: IGameData) => {
      const msg = 'Game Restarted Due to Settings Change 😅';
      toast.current?.show({ severity: 'info', summary: 'Settings Changed', detail: msg, life: 3000 });
      dispatch(initializeGame(gameData));
    })
  }

  const destroyListeners = (): void => {
    socket.off(EVENTS.GAME_CHECK_SUCCESS);
    socket.off(EVENTS.GAME_CHECK_FAILED);
    socket.off(EVENTS.GAME_RESTART_AGAIN);
    socket.off(EVENTS.SETTINGS_CHANGED);
    socket.off(EVENTS.GAME_UPDATE);
  }


  const isUserNameSet = (): boolean => {
    return playerName === '' ? false : true;
  }

  const setGameIfExist = (): void => {
    socket.emit(EVENTS.GAME_CHECK, gameId)
  }

  const getWinLocationIfExists = (row: number, col: number, grid: GridCol[][]): number[][] => {
    const combinations = GameUtils.generatePossibleWinCombinations(row, col, gridDimension);
    console.log('combinations ',combinations, gridDimension)
    let location: number[][] = []
    combinations.forEach(combination => {
      let combinationStr = ''
      let tempLocations:number[][]= []
      combination.forEach((location)=>{
        const [row, col] = location.split('');
        combinationStr += `${grid[+row][+col]}`;
        tempLocations.push([+row, +col])
      })
      const isWin = [E_GAMECHARACTER.X.repeat(gridDimension), E_GAMECHARACTER.O.repeat(gridDimension)].includes(combinationStr);
      console.log(E_GAMECHARACTER.X.repeat(gridDimension),combinationStr)
      if (isWin) location.push(...tempLocations);
    })
    return location;
  }

  const canPlay = () => {
    if (playerCharacter === 'X') {
      return playCount % 2 === 0;
    }
    else {
      return playCount % 2 !== 0;
    }
  }

  const playTurn = (row: number, col: number): void => {
    if (canPlay()) {
      if (game.grid[row][col] !== null) {
        const message = `Can't play in this square! pick an empty square to play in`;
        toast.current?.show({ severity: 'info', summary: 'Wait your turn', detail: message, life: 3000 });
        return;
      }
      const newGrid = GameUtils.deepClone<GridCol[][]>(game.grid);
      newGrid[row][col] = playerCharacter;
      // playSound.play();
      const gameUpdateData: IUpdateGameData = {
        roomId: gameId,
        playCount: playCount + 1,
        playLocation: [row, col]
      }
      socket.emit(EVENTS.GAME_UPDATE, gameUpdateData);
    }
    else {
      const message = `It's ${friendName} turn to play, please wait for your turn`;
      toast.current?.show({ severity: 'info', summary: 'Illegal Play', detail: message, life: 3000 })
    }
  }

  const StatusIcon = {
    Win: '🏆',
    win2: '🎉',
    Draw: '⚖',
    InPlay: '🕓'
  }

  const renderCharacterColor = (value: string): string => {
    return value === 'X' ? orange[100] : orange[200]
  }

  const isPlayLocation = (row: number, col: number): boolean => {
    let currentLocation = currentPlayLocation.length > 0 ? `${currentPlayLocation[0]}${currentPlayLocation[1]}` : '';
    return currentLocation === `${row}${col}`
  }

  const updateStatusMessage = (status: GameStatus, winCharacter: string = ''): string => {
    let statusMsg = '';
    switch (status) {
      case GameStatus.InPlay: {
        statusMsg = canPlay() ? `It's your Turn` : `It's ${friendName}'s Turn`;
        break;
      }
      case GameStatus.Win: {
        statusMsg = `${winCharacter === playerCharacter ? 'You' : friendName} win :)`;
        break;
      }
      case GameStatus.Draw: {
        statusMsg = `This Game is a Tie`;
        break;
      }
      default: { }
    }
    return statusMsg;
  }

  const isWin = (): boolean => game.winLocation.size > 0;
  const isDraw = (): boolean => playCount === Math.pow(gridDimension,2) && !isWin();
  const playAgain = (): void => {
    if (gameId) {
      const data: IPlayAgainData = {
        roomId: gameId,
        isDraw: isDraw(),
        winCharacter: game.winCharacter
      }
      socket.emit(EVENTS.PLAY_AGAIN, data)
    }
  }
  const isWinLocation = (row: number, col: number): boolean => {
    return game.winLocation.has(`${row}${col}`)
  }
  const startCountDown = () => {
    const interval: any = setInterval(() => {
      setSeconds((prevState) => prevState - 1)
    }, 1000)
    setIntervalId(interval)
  }
  const stopCountDown = () => {
    clearInterval(intervalId);
    setSeconds(59);
  }
  const toggleFullScreen = ()=>{
    if (!document.fullscreenElement) {
      document.documentElement.requestFullscreen();
    } else if (document.exitFullscreen) {
      document.exitFullscreen();
    }
  }
  return {
    gameReady,
    wins,
    losses,
    draws,
    game,
    playerImage,
    playerNo,
    playerName,
    playerCharacter,
    friendImage,
    friendNo,
    friendName,
    friendCharacter,
    playCount,
    seconds,
    gameMode,
    StatusIcon,
    gridDimension,
    isFullscreen,
    renderCharacterColor,
    canPlay,
    isWin,
    isDraw,
    isPlayLocation,
    isWinLocation,
    playTurn,
    playAgain,
    toggleFullScreen
  }
}

export default useGameHooks;
