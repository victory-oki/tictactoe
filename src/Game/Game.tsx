import React, { useState } from 'react'
import { GameContainerStyled } from './GameStyles';
import { ReactComponent as DrawIcon } from '../assets/draw-icon.svg';
import { ProgressSpinner } from 'primereact/progressspinner';
import { orange } from '../theme';
import Avatar from '../components/Avatar/Avatar';
import { Button } from 'primereact/button';
import useGameHooks from './useGameHooks';
import { E_GAMECHARACTER, E_GAMEMODE, IGameHooks } from '../gameModel';
import SoundPlayer from '../components/SoundPlayer/SoundPlayer';
import Settings from '../components/Settings/Settings';
import { Tooltip } from 'primereact/tooltip';

function Game() {
  const {
    gameReady,
    wins,
    losses,
    draws,
    game,
    playerImage,
    playerNo,
    playerName,
    playerCharacter,
    friendImage,
    friendNo,
    friendName,
    friendCharacter,
    seconds,
    gameMode,
    StatusIcon,
    gridDimension,
    isFullscreen,
    playCount,
    renderCharacterColor,
    canPlay,
    isWin,
    isDraw,
    isPlayLocation,
    isWinLocation,
    playTurn,
    playAgain,
    toggleFullScreen
  }: IGameHooks = useGameHooks();

  const [isOpened, openModal] = useState(false);
  return (
    <GameContainerStyled>
      <Settings isOpened={isOpened} openModal={openModal} />
      <div className="container">
        {
          !gameReady ? (
            <>
              <ProgressSpinner style={{ width: '150px', height: '150px' }} strokeWidth="6" fill="transparent" animationDuration=".5s" />
            </>
          ) : (
            <>
              <div className="nav">
                <p className="title">Crosses & Noughts</p>
                <ul className="list">
                  <li className="list__item">
                    {
                      (!isWin() && !isDraw() && playCount > 0) && <Tooltip target=".disabled-button" position={'bottom'} />
                    }
                    <span className="disabled-button mr-2" data-pr-tooltip="Can't change Settings While a game is in play ðŸ˜…">
                      <i className="pi nav__icon pi-cog" onClick={() => {
                        if (isWin() || isDraw() || playCount < 1) {
                          openModal(true);
                        }
                      }}></i>
                    </span>
                  </li>
                  <li className="list__item"><SoundPlayer url={require(`../assets/sound.mp3`)} config={{ autoPlay: false, loop: true }} /></li>
                  <li className="list__item" onClick={()=>toggleFullScreen()}>{
                  isFullscreen ? <i className='pi nav__icon pi-window-minimize' />: <i className='pi nav__icon pi-window-maximize' />
                  }</li>
                </ul>
              </div>
              <div className="header">
              </div>
              <div className="status">
                <div className="status__icon">{StatusIcon[game.gameStatus]}</div>
                <div className="status__text">
                  <p className="status__text--heading">Status</p>
                  <div className="status__footer">
                    <p className="status__text--content">{game.msgStatus}</p>
                    {
                      isWin() || isDraw() ? (
                        <Button label="Play Again" className="p-button-outlined p-button-secondary status__btn" onClick={
                          () => playAgain()
                        } />
                      ) : null
                    }
                    {
                      gameMode === E_GAMEMODE.BLITZ && canPlay() && !isWin() && !isDraw() ? (
                        <p
                          className={'time-label ' + (seconds < 30 && seconds > 10 ? 'time-label--warning ' : '') +
                            (seconds <= 10 ? 'time-label--danger animate__animated animate__pulse animate__infinite' : '')}>
                          00:{`${seconds}`.length > 1 ? `${seconds}` : `0${seconds}`}
                        </p>
                      ) : null
                    }
                  </div>
                </div>
              </div>
              <div className="game">
                <div className="player">
                  <div className="player__img">
                    <Avatar fileName={`${playerImage}.svg`} />
                    <i className={canPlay() && !isWin() && !isDraw() ? 'pi pi-angle-down animate__animated animate__bounce animate__infinite show' : ''}></i>
                    <p className="player__char" style={{ color: renderCharacterColor(playerCharacter) }}>{playerCharacter}</p>
                  </div>
                  <p className="player__id">Player {playerNo}</p>
                  <p className="player__name">{playerName}</p>
                  <p className="player__stats">
                    <span className="player_stats--win" style={{ color: renderCharacterColor(playerCharacter) }}>{wins} wins</span>
                    <span className="player_stats--draw">{draws} draws</span>
                  </p>
                </div>
                <div className="grid">
                  {
                    game.grid.map((item: any[], rowIndex: number) => (
                      <div className="row" key={rowIndex}>
                        {
                          item.map((colItem, colIndex) => (
                            <div
                              className="col"
                              onClick={() => !isWin() && !isDraw() && playTurn(rowIndex, colIndex)}
                              key={colIndex}
                              style={{
                                color: renderCharacterColor(colItem ?? ''),
                                background: isWinLocation(rowIndex, colIndex) ? orange[50] : 'hsla(0, 0%, 98%, 1)'
                              }}
                            >
                              <p className={isPlayLocation(rowIndex, colIndex) ? 'animate__animated animate__heartBeat animate__repeat-1' : ''}>
                                {colItem ?? ''}
                              </p>
                            </div>
                          ))
                        }
                      </div>
                    ))
                  }
                </div>
                <div className="player">
                  <div className="player__img">
                    <Avatar fileName={`${friendImage}.svg`} />
                    <i className={!canPlay() && !isWin() && !isDraw() ? 'pi pi-angle-down animate__animated animate__bounce animate__infinite show' : ''}></i>
                    <p className="player__char player__char--right" style={{ color: renderCharacterColor(friendCharacter) }}>{friendCharacter}</p>
                  </div>
                  <p className="player__id">Player {friendNo}</p>
                  <p className="player__name">{friendName}</p>
                  <p className="player__stats">
                    <span className="player_stats--win" style={{ color: renderCharacterColor(friendCharacter) }}>{losses} wins</span>
                    <span className="player_stats--draw">{draws} draws</span>
                  </p>
                </div>
              </div>
            </>
          )
        }
      </div>
    </GameContainerStyled>
  )
}

export default Game

