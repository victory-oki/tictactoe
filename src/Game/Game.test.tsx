import React from 'react';
import { describe, expect, test } from '@jest/globals';
import * as hooks from './useGameHooks';
import Game from './Game';
import { E_GAMEMODE, GameStatus, GridCol } from '../gameModel';
import { testSnapshot, wrapContext } from '../testHelpers';

const gridInit: GridCol[][] = [[null, null, null], [null, null, null], [null, null, null]];
const defaultMockreturnValue = {
    gameReady: false,
    wins: 0,
    losses: 0,
    draws: 0,
    game:{
        gameStatus: GameStatus.InPlay,
        msgStatus: '',
        winLocation:new Set<string>(),
        grid: gridInit,
        winCharacter: ''
    },
    playerName: 'Tory',
    playerNo: 1,
    friendNo: 2,
    friendName: 'Pat',
    playerCharacter: 'X',
    friendCharacter: 'O',
    playerImage: 1,
    friendImage: 2,
    StatusIcon:{
        Win: '🏆',
        win2: '🎉',
        Draw: '⚖',
        InPlay: '🕓'
    },
    seconds:59,
    gameMode:E_GAMEMODE.NORMAL,
    renderCharacterColor: jest.fn(),
    canPlay: jest.fn(),
    isWin: jest.fn(),
    isDraw: jest.fn(),
    isPlayLocation: jest.fn(),
    isWinLocation: jest.fn(),
    playTurn: jest.fn(),
    playAgain: jest.fn()
};

describe('Game', () => {
    const wrappedComponent: JSX.Element = wrapContext(<Game/>);
    it('should render the component in loading state', () => {
        const hookSpy = jest.spyOn(hooks, 'default').mockReturnValue(defaultMockreturnValue);

        testSnapshot(wrappedComponent);

        expect(hookSpy).toHaveBeenCalledTimes(1);

    });

    it('should render the component in default state', () => {
        const hookSpy = jest.spyOn(hooks, 'default').mockReturnValue({...defaultMockreturnValue, gameReady: true});

        testSnapshot(wrappedComponent, true);
    });

    // test('generate Possible Win Combinations (no diagonals)', () => {
    //     const arrayIsSame = (array1: string[], array2: string[]) => {
    //         const set = new Set(array1);
    //         return array2.every(val => set.has(val));
    //     };
    // });
});
