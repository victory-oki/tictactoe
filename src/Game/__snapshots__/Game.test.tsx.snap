// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render the component in default state 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Game />
  </Provider>
</Router>
`;

exports[`Game should render the component in loading state 1`] = `
<div
  className="sc-gsnTZi knoPkv"
>
  <div
    className="container"
  >
    <div
      aria-busy={true}
      className="p-progress-spinner"
      id={null}
      role="alert"
      style={
        Object {
          "height": "150px",
          "width": "150px",
        }
      }
    >
      <svg
        className="p-progress-spinner-svg"
        style={
          Object {
            "animationDuration": ".5s",
          }
        }
        viewBox="25 25 50 50"
      >
        <circle
          className="p-progress-spinner-circle"
          cx="50"
          cy="50"
          fill="transparent"
          r="20"
          strokeMiterlimit="10"
          strokeWidth="6"
        />
      </svg>
    </div>
  </div>
</div>
`;
